Review decorator vignette (#1470)
insightsengineering/teal: vignettes/customizing-module-output.Rmd

---
title: "Customizing Module Output"
author: "NEST CoreDev"
output:
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Customizing Module Output}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction

`teal` provides a collection of ready-to-use modules, each with a predefined way of presenting outputs.

While these built-in modules cover many common use cases, you may need to customize their outputs to better fit your specific requirements.
This document outlines the available customization options for modifying `teal` module outputs.

You will learn how to use `teal_transform_module()` to modify and enhance the objects created by `teal::modules()`,
allowing you to tailor the outputs without rewriting the original module code.

## Decorators

In `teal`, decorators are used to modify module outputs, such as plots or tables, without changing their core structure.

> For example, imagine you have a plot and you want to add a title.
> A decorator function, like `add_title(x, "My Plot")`, takes an existing plot and updates it by adding a title while keeping everything else the same.
> This makes it different from a function that creates a brand-new plot from scratch, as decorators work by enhancing what’s already there.

Reviewer: ```suggestion
For example, imagine you have a `plot` object (ggplot) and you want to add a title. To do so one simply needs to call `plot <- plot + ggtitle("My Plot")`. It is important to enhance the current object object instead of creating a new one. We call this operation a decoration.
```

The decorators process can vary in complexity:

- **Simple Decorators**: Single-step modifications, such as a single method call that does not require additional data.
- **Complex Decorators**: Multi-step operations that may involve interdependent transformations, potentially requiring input from dedicated `shiny` UI elements.

This functionality empowers application developers to significantly customize outputs beyond the default capabilities provided by existing module parameters.

> Decorators allow for advanced modifications, enabling highly tailored and dynamic user experiences in `teal` applications.

Reviewer: I don't understand what these two sentences means. They don't help app developer grasp what is decoration. Besides these statements doesn't have to be true because:

> This functionality empowers application developers to significantly customize outputs beyond the default capabilities provided by existing module parameters

Signification is debatable. For example one can't change object type from plot to for example table. Also decorator is "existing module parameter" so it doesn't go beyond "existing module parameters" - it goes beyond default output appearance.

Author: These two sentences feel like closing statements to me. 

However, since we already have this sentence at the beginning of the section:
```
In `teal`, decorators are used to modify module outputs, such as plots or tables, without changing their core structure.
```  

I think we can remove these two sentences without losing clarity.

WDYT?

Reviewer: Yes, I think also to remove.

## Requirements and Limitations

To use decorators effectively, certain requirements must be met:

1. **Module Support**: While `teal` provides the core functionality for decorators, the module must explicitly support this functionality. Developers should ensure that the module has been designed to work with decorators (see [Include Decorators in a `teal` Module](#include-decorators-in-a-teal-module)).

> 2. **Matching Object Names**: Decorators must reference object names that align with the internal naming conventions of the module. Each module may use different names for its output objects, such as `plot` or `table`. This alignment is critical for successful decorator.

Reviewer: ```suggestion
2. **Matching Object Names**: Decorators must reference object names that align with the internal naming conventions of the module. Each module may use different names for its output objects, such as `plot` or `table`. This alignment is critical for successful decorator.
3. **Matching Object Class**: Decorated objects are used in certain context and they are use to be consumed by relevant `render*` functions. For example `ggplot` objects are used in `renderPlot` while `plotly` objects are consumed by `renderPlotly`. It is important that the object doesn't change its basic class so the module can deal with the modified object.     
```

> It is recommended to review the module documentation or source code to understand its internal object naming before applying decorators.

Reviewer: ```suggestion
It is recommended to review the module documentation or source code to understand its internal object naming and object class before applying decorators.
```

---
title: "Customizing Module Output"
author: "NEST CoreDev"
output:
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Customizing Module Output}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction

> `teal` is a powerful `shiny`-based framework with built-in modules for interactive data analysis.

Reviewer: I'd skip this line at all. This is decorator vignette and everyone should already know what is teal. Also "powerful" is a bragging which shouldn't really have a place in the vignette. (this is just my opinion)

```suggestion
```

Author: Agree.
